### PR TITLE
ENH - fit intercept in ``ProxNewton`` solver

### LINKED ISSUE #68: ENH fit intercept in prox newton solver

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
This adds unpenalized intercept support to ``ProxNewton`` solver and closes #68.
It proceeds as follows
- [x] implement fit intercept
- [x] add unitest
- [x] perform benchmarks

-------------------
[View benchmark results](https://badr-moufad.github.io/share-benchmarks/pn-intercept/benchmark_logreg_l1_benchopt_run_2022-10-04_08h33m33.html)
[Link to benchmark repository](https://github.com/Badr-MOUFAD/benchmark_logreg_l1/tree/bench-intercept)

### CODE REVIEW

File: skglm/solvers/prox_newton.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ class ProxNewton(BaseSolver):
     """
 
     def __init__(self, p0=10, max_iter=20, max_pn_iter=1000, tol=1e-4,
-                 fit_intercept=True, warm_start=False, verbose=0):
+                 fit_intercept=False, warm_start=False, verbose=0):
```
Comment (mathurinm) [inline skglm/solvers/prox_newton.py:45]: changing default has a silent impact on users, we try to avoid it as much as possible (though our userbase may not be so large, let's take good habits from the start)

File: skglm/tests/test_prox_newton.py [hunk 2]
```diff
@@ -20,13 +20,13 @@ def test_alpha_max(X_density):
```
Comment (mathurinm) [inline skglm/tests/test_prox_newton.py:20]: alpha_max changes when fitting intercept

File: skglm/tests/test_prox_newton.py [hunk 3]
```diff
@@ -37,14 +37,17 @@ def test_pn_vs_sklearn(rho, X_density):
     alpha = rho * alpha_max
 
     sk_log_reg = LogisticRegression(penalty='l1', C=1/(n_samples * alpha),
-                                    fit_intercept=False, tol=1e-9, solver='liblinear')
+                                    fit_intercept=fit_intercept,
```
Comment (mathurinm) [inline skglm/tests/test_prox_newton.py:40]: liblinear penalizes intercept IIRC. Does it work fine here?

File: skglm/tests/test_prox_newton.py [hunk 4]
```diff
@@ -37,14 +37,17 @@ def test_pn_vs_sklearn(rho, X_density):
...
 
     log_datafit = compiled_clone(Logistic())
     l1_penalty = compiled_clone(L1(alpha))
-    w = ProxNewton(tol=1e-9).solve(X, y, log_datafit, l1_penalty)[0]
+    prox_solver = ProxNewton(fit_intercept=fit_intercept, tol=1e-9)
+    w = prox_solver.solve(X, y, log_datafit, l1_penalty)[0]
 
-    np.testing.assert_allclose(w, sk_log_reg.coef_.flatten(), rtol=1e-6, atol=1e-6)
+    np.testing.assert_allclose(w[:n_features], sk_log_reg.coef_.flatten(),
+                               rtol=1e-6, atol=1e-6)
```
Comment (mathurinm) [inline skglm/tests/test_prox_newton.py:50]: if fit intercept we can also check that the intercepts are non zero and equal

File: skglm/solvers/prox_newton.py (hunk 1 above)
Comment (Badr-MOUFAD) [reply skglm/solvers/prox_newton.py:45]: Yes, I totally agree with you. I will revert it.
(I made such a change to pass tests)

Comment (mathurinm) [reply skglm/solvers/prox_newton.py:45]: perfect :)

File: skglm/tests/test_prox_newton.py (hunk 2 above)
Comment (Badr-MOUFAD) [reply skglm/tests/test_prox_newton.py:20]: It is ``alpha_max`` for the penalized coefficients. Hence adding an unpenalized intercept doesn't change it.

If I understand what you mean: the condition to have a ``0`` intercept translates to having a centered ``y``.

File: skglm/tests/test_prox_newton.py (hunk 3 above)
Comment (Badr-MOUFAD) [reply skglm/tests/test_prox_newton.py:40]: No it doesn't work here. The test against ``sklearn`` fails.
I was actually scratching my head to figure out why. Thanks @mathurinm for reminding me that liblinear fits an unpenalized intercept.

Any ideas against whom should we uni test?

File: skglm/tests/test_prox_newton.py (hunk 2 above)
Comment (mathurinm) [reply skglm/tests/test_prox_newton.py:20]: That would be the case for the Lasso, but for Logreg things are different. 
My point is, you compute the alpha_max without intercept, then fit with an intercept, so without more computations, it's not clear what should happen (should we expect 0 as a solution or not?)

File: skglm/tests/test_prox_newton.py (hunk 3 above)
Comment (mathurinm) [reply skglm/tests/test_prox_newton.py:40]: as a first try i'd test `intercept_scaling=1000`  in sklearn logreg it it doesn't break convergence (lower value else)

Comment (QB3) [reply skglm/tests/test_prox_newton.py:40]: > liblinear fits an unpenalized intercept

Did you mean 'liblinear fits a penalized intercept' ? https://github.com/scikit-learn/scikit-learn/issues/10626

Comment (QB3) [reply skglm/tests/test_prox_newton.py:40]: > Any ideas against whom should we uni test?

Could we compare to our own cd solvers with intercept ?

Comment (mathurinm) [reply skglm/tests/test_prox_newton.py:40]: yes penalized, my bad
just try a high intercept scaling first to see if you get a value within reasonable tolerance of our unpenalized intercept

Comment (Badr-MOUFAD) [reply skglm/tests/test_prox_newton.py]: I checked against [``scikit-learn:saga``](https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression:~:text=The%20following%20table%20summarizes%20the%20penalties%20supported%20by%20each%20solver%3A) as it fits unpenalized intercept

File: skglm/tests/test_prox_newton.py (hunk 2 above)
Comment (Badr-MOUFAD) [reply skglm/tests/test_prox_newton.py:20]: I 100% agree with you! Yet, in terms of ``alpha_max``, Lasso resemble slightly LogReg as $\nabla F(0) = \frac{y}{2 n}$

$$ -\mathbf{X}^\top \nabla F(0) \in \lambda \partial \lVert \cdot \rVert_1 \quad \mathbb{1}_n^\top \nabla F(0) = 0$$

$$   \lVert \frac{\mathbf{X}^\top y}{2 n}  \rVert_\infty \leq \lambda, \quad \bar{y} = 0$$

I have added a comment in the unittest to clarify that.

**Comment by Badr-MOUFAD** [comment]
here are the [benchmark results](https://badr-moufad.github.io/share-benchmarks/pn-intercept/benchmark_logreg_l1_benchopt_run_2022-10-04_08h33m33.html) and the [link to repo](https://github.com/Badr-MOUFAD/benchmark_logreg_l1/tree/bench-intercept) to reproduce.

I have one concern: I cannot explain why we don't have the same initial objective value with Blitz in the case of ``alpha=alpha_max`` (``reg=1``)

any thoughts?

**Comment by mathurinm** [comment]
I'd test the output (intercept and coef_) outside of benchopt for these two solvers on the problematic data. Check that both get 0 (but I don't think they should, cf our discussion on alpha_max with intercept)

File: skglm/tests/test_prox_newton.py (hunk 2 above)
Comment (Badr-MOUFAD) [reply skglm/tests/test_prox_newton.py:20]: I have re-checked. There are some subtleties in the intercept case.
I resolved to exclude ``fit_intercept=True`` from the ``alpha_max`` unit test.

What do you think @mathurinm?

**Comment by Badr-MOUFAD** [comment]
> I'd test the output (intercept and coef_) outside of benchopt for these two solvers on the problematic data. Check that both get 0 (but I don't think they should, cf our discussion on alpha_max with intercept)

The reason behind this is: Blitz performs an intercept update before entering the ``iter`` loop.
cf. https://github.com/tbjohns/BlitzL1/blob/8c1d2f6251307021ff4c8e2027df2d92b10d7f63/src/solver.cpp#L442-L443

So nothing surprising in such behavior.

File: skglm/tests/test_prox_newton.py (hunk 2 above)
Comment (Badr-MOUFAD) [reply skglm/tests/test_prox_newton.py:20]: I don't think it's a pathological case as ``alpha_max`` changes when fitting an unpenalized intercept.

I can show that if we set ``alpha_max = norm(X.T @ y, ord=np.inf) / n_samples`` (remove 2 from denominator),
we would have ``w[:n_features] == 0`` in both cases (``fit_intercept`` ``True`` or ``False``), 

However, I don't think it's a clever idea to refactor this unittest to handle ``fit_intercept=True`` and write a long math justification to explain it. Indeed, we have a second unittest that checks thoroughly the case ``fit_intercept=True``

Comment (mathurinm) [reply skglm/tests/test_prox_newton.py:20]: I agree, it is not worth the complication. Tests should make our life easier not complicate the codebase :)

File: skglm/tests/test_prox_newton.py [hunk 5]
```diff
@@ -9,8 +9,8 @@
 from skglm.utils import make_correlated_data, compiled_clone
 
 
-@pytest.mark.parametrize('X_density', [1, 0.5])
-def test_alpha_max(X_density):
+@pytest.mark.parametrize('X_density, fit_intercept', product([1, 0.5], [False]))
+def test_alpha_max(X_density, fit_intercept):
```
Comment (mathurinm) [inline skglm/tests/test_prox_newton.py:13]: remove fit_intercept from this test

File: skglm/solvers/prox_newton.py [hunk 6]
```diff
@@ -63,20 +67,37 @@ def solve(self, X, y, datafit, penalty, w_init=None, Xw_init=None):
...
             else:
-                grad = _construct_grad(X, y, w, Xw, datafit, all_features)
+                grad = _construct_grad(X, y, w[:n_features], Xw, datafit, all_features)
 
-            opt = penalty.subdiff_distance(w, grad, all_features)
+            opt = penalty.subdiff_distance(w[:n_features], grad, all_features)
+
+            # optimality of intercept
+            if fit_intercept:
+                intercept_opt = np.absolute(np.sum(datafit.raw_grad(y, Xw)))
```
Comment (PABannier) [inline skglm/solvers/prox_newton.py:93]:
```suggestion
                intercept_opt = np.abs(np.sum(datafit.raw_grad(y, Xw)))
```

**Review by PABannier — APPROVED** [review]
LGTM! Really good work @Badr-MOUFAD thanks!

File: skglm/solvers/prox_newton.py [hunk 7]
```diff
@@ -24,6 +24,9 @@ class ProxNewton(BaseSolver):
     tol : float, default 1e-4
         Tolerance for convergence.
 
+    fit_intercept : bool, default True
```
Comment (mathurinm) [inline skglm/solvers/prox_newton.py:27]: My experience with sklearn is that it's quite confusing to fit an intercept by default. Should it be set to False here ?

File: skglm/solvers/prox_newton.py [hunk 8]
```diff
@@ -63,20 +67,37 @@ def solve(self, X, y, datafit, penalty, w_init=None, Xw_init=None):
...
             else:
-                grad = _construct_grad(X, y, w, Xw, datafit, all_features)
+                grad = _construct_grad(X, y, w[:n_features], Xw, datafit, all_features)
 
-            opt = penalty.subdiff_distance(w, grad, all_features)
+            opt = penalty.subdiff_distance(w[:n_features], grad, all_features)
+
+            # optimality of intercept
+            if fit_intercept:
+                intercept_opt = np.abs(np.sum(datafit.raw_grad(y, Xw)))
```
Comment (mathurinm) [inline skglm/solvers/prox_newton.py:93]: To be sure I understand correctly : summing is like taking dot product with np.ones(n_features), so that's the magnitude of the gradient with respect to a constant feature here ?

Can you add a comment ("gradient wrt intercept" or "gradient with respect to constant feature of 1s") ?

File: skglm/solvers/prox_newton.py [hunk 9]
```diff
@@ -174,22 +202,35 @@ def _descent_direction(X, y, w_epoch, Xw_epoch, grad_ws, datafit,
...
+
+            if stop_crit <= tol:
                 break
 
     # descent direction
-    return w_ws - w_epoch[ws], X_delta_w_ws
+    return w_ws - w_epoch[ws_intercept], X_delta_w_ws
 
 
 # sparse version of _compute_descent_direction
```
Comment (mathurinm) [inline skglm/solvers/prox_newton.py:230]:
```suggestion
# sparse version of _descent_direction
```

File: skglm/tests/test_prox_newton.py [hunk 10]
```diff
@@ -20,14 +20,15 @@ def test_alpha_max(X_density):
...
+    w = ProxNewton(fit_intercept=False).solve(X, y, log_datafit, l1_penalty)[0]
 
     np.testing.assert_equal(w, 0)
 
 
-@pytest.mark.parametrize("rho, X_density", product([1e-1, 1e-2], [1., 0.5]))
-def test_pn_vs_sklearn(rho, X_density):
-    n_samples, n_features = 11, 19
+@pytest.mark.parametrize("rho, X_density, fit_intercept",
+                         product([1e-1, 1e-2], [1., 0.5], [True, False]))
```
Comment (mathurinm) [inline skglm/tests/test_prox_newton.py:29]: @Badr-MOUFAD is there a particular thing to test rho=1 ? For me if it works for rho=0.1 it works for rho=1. We don't test anything particular for rho=1.

If not, it can be removed (the test will be simpler and have the same coverage)

File: skglm/tests/test_prox_newton.py [hunk 11]
```diff
@@ -37,14 +38,19 @@ def test_pn_vs_sklearn(rho, X_density):
...
     l1_penalty = compiled_clone(L1(alpha))
-    w = ProxNewton(tol=1e-9).solve(X, y, log_datafit, l1_penalty)[0]
+    prox_solver = ProxNewton(fit_intercept=fit_intercept, tol=1e-9)
+    w = prox_solver.solve(X, y, log_datafit, l1_penalty)[0]
 
-    np.testing.assert_allclose(w, sk_log_reg.coef_.flatten(), rtol=1e-6, atol=1e-6)
+    np.testing.assert_allclose(w[:n_features], sk_log_reg.coef_.flatten(),
+                               rtol=1e-4, atol=1e-3)
+    if fit_intercept:
+        np.testing.assert_allclose(w[-1], sk_log_reg.intercept_, rtol=1e-4, atol=1e-3)
```
Comment (mathurinm) [inline skglm/tests/test_prox_newton.py:53]: @Badr-MOUFAD use atol and rtol only when needed. The tests seem to pass even when they take their default values here. If not, consider lowering the tol.

atol=1e-3 may be big in some settings, so better avoid it.

File: skglm/solvers/prox_newton.py (hunk 7 above)
Comment (Badr-MOUFAD) [reply skglm/solvers/prox_newton.py:27]: agree!

File: skglm/tests/test_prox_newton.py (hunk 10 above)
Comment (Badr-MOUFAD) [reply skglm/tests/test_prox_newton.py:29]: I have no strong opinion.
It's just a sanity check that we've used to implement.

Comment (mathurinm) [reply skglm/tests/test_prox_newton.py:29]: Then I have a strong opinion :) 
It's a sanity check when we use benchopt, to check that the objective doesn't move. Here, we don't check that, so this parameter can be removed to make the test simpler. Just use a hardcoded rho=0.1

File: skglm/tests/test_prox_newton.py (hunk 11 above)
Comment (mathurinm) [reply skglm/tests/test_prox_newton.py:52]: spoke too fast, some tests seem to break. Can you check if lowering the tolerance of solvers allows us to remove the rtol and atol ? Or just to use atol=1e-9 ?

**Review by mathurinm — APPROVED** [review]

**Comment by mathurinm** [comment]
Thanks @Badr-MOUFAD !
